qiskit/quantum_info/mitigation/complete_readout_mitigator.py

```diff
@@ -0,0 +1,281 @@
+        """
+        # Marginalize counts
+        if clbits is not None:
+            data = marginal_counts(data, clbits)
+
+        # Get probability vector
+        num_qubits = data.size()
+        probs_vec = self._to_probs_vec(data, num_qubits)
+        shots = data.shots()
+
+        # Get qubit mitigation matrix and mitigate probs
+        if qubits is None:
+            qubits = range(num_qubits)
+        mit_mat = self.mitigation_matrix(qubits)
+
+        # Get operator coeffs
+        if diagonal is None:
+            diagonal = self._z_diagonal(2 ** num_qubits)
+        else:
+            diagonal = self._str2diag(diagonal)
+
+        # Apply transpose of mitigation matrix
+        coeffs = mit_mat.T.dot(diagonal)
+
+        return self._expval_with_stddev(coeffs, probs_vec, shots)
+
+    def quasi_probabilities(
+        self,
+        data: Counts,
+        qubits: Optional[List[int]] = None,
+        clbits: Optional[List[int]] = None,
+        num_qubits: Optional[int] = None,
+        shots: Optional[bool] = False,
+    ) -> (QuasiDistribution, Dict[str, float]):
+        """Compute mitigated quasi probabilities value.
+
+        Args:
+            data: counts object
+            qubits: qubits the count bitstrings correspond to.
+            clbits: Optional, marginalize counts to just these bits.
+            num_qubits: the total number of qubits.
+            shots: return the number of shots.
+
+        Raises:
+            QiskitError: if qubit and clbit kwargs are not valid.
+        """
+        # Marginalize counts
+        if clbits is not None:
+            data = marginal_counts(data, clbits)
+
+        # Get total number of qubits and shots
+        num_qubits = data.size()
+        shots = data.shots()
+        # Get probability vector
+        probs_vec = self._to_probs_vec(data, num_qubits)
+
+        # Get qubit mitigation matrix and mitigate probs
+        if qubits is None:
+            qubits = range(num_qubits)
+        mit_mat = self.mitigation_matrix(qubits)
+
+        # Apply transpose of mitigation matrix
+        probs_vec = mit_mat.dot(probs_vec)
+        probs_dict = {}
+        for index in range(len(probs_vec)):
+            probs_dict[index] = probs_vec[index]
+
+        return QuasiDistribution(probs_dict), self._stddev(probs_dict, shots)
+
+    def mitigation_matrix(self, qubits: List[int] = None) -> np.ndarray:
+        r"""Return the readout mitigation matrix for the specified qubits.
+        The mitigation matrix :math:`A^{-1}` is defined as the inverse of the
+        :meth:`assignment_matrix` :math:`A`.
+        Args:
+            qubits: Optional, qubits being measured.
+        Returns:
+            np.ndarray: the measurement error mitigation matrix :math:`A^{-1}`.
+        """
+        if qubits is None:
+            qubits = tuple(range(self._num_qubits))
+        else:
+            qubits = tuple(sorted(qubits))
+
+        # Check for cached mitigation matrix
```

Comment:
Why not use a built-in Python caching mechanism such as the `@lru_cache` decorator?